samples/RefreshFlow/AngularApp/src/core/auth.service.ts

```diff
@@ -0,0 +1,164 @@
+
+        this.loggedIn$ = this.tokens$.map(tokens => !!tokens);
+
+        return this.startupTokenRefresh()
+            .do(() => this.scheduleRefresh());
+    }
+
+    register(data: RegisterModel): Observable<Response> {
+        return this.http.post('http://localhost:5056/account/register', data)
+            .catch(res => Observable.throw(res.json()));
+    }
+
+    login(user: LoginModel) {
+        return this.getTokens(user, 'password')
+            .catch(res => Observable.throw(res.json()))
+            .do(res => this.scheduleRefresh());
+    }
+
+    logout() {
+        this.updateState({ profile: null, tokens: null });
+        if (this.refreshSubscription$) {
+            this.refreshSubscription$.unsubscribe();
+        }
+        this.removeToken();
+    }
+
+    isInRole(usersRole: string): Observable<boolean> {
+        return this.profile$
+            .map(profile => profile.role.find(role => role === usersRole) !== undefined);
+    }
+
+    refreshTokens() {
+        return this.state.first()
+            .map(state => state.tokens)
+            .flatMap(tokens => this.getTokens({ refresh_token: tokens.refresh_token }, 'refresh_token')
+                .catch(error => Observable.throw('Session Expired'))
+            );
+    }
+
+    private storeToken(tokens: AuthTokenModel) {
+        localStorage.setItem('auth-tokens', JSON.stringify(tokens));
+    }
+    private retrieveTokens() {
+        const tokensString = localStorage.getItem('auth-tokens');
+        const tokensModel: AuthTokenModel = tokensString == null ? null : JSON.parse(tokensString);
+        return tokensModel;
+    }
+    private removeToken() {
+        localStorage.removeItem('auth-tokens');
+    }
+
+    private updateState(newState: AuthStateModel) {
+        const previoudState = this.state.getValue();
+        this.state.next(Object.assign({}, previoudState, newState));
+    }
+
+    private getTokens(data: RefreshGrantModel | LoginModel, grantType: string) {
+        const headers = new Headers({ 'Content-Type': 'application/x-www-form-urlencoded' });
+        const options = new RequestOptions({ headers: headers });
+
+        Object.assign(data, { grant_type: grantType, scope: 'openid offline_access' });
+
+        const params = new URLSearchParams();
+        Object.keys(data).forEach(key => params.append(key, (<any>data)[key]));
+
+        return this.http.post('http://localhost:5056/connect/token', params.toString(), options)
+            .map(res => {
+                const tokens: AuthTokenModel = res.json();
+                const now = new Date();
+                tokens.expiration_date = new Date(now.getTime() + tokens.expires_in * 1000).getTime().toString();
+
+                const profile: ProfileModel = jwtDecode(tokens.id_token);
+
+                this.storeToken(tokens);
+                this.updateState({ authReady: true, tokens, profile });
+                return res;
+            });
+    }
```

Comment:
This is an example of inconsistent vertical whitespace. I would expect a consistent use of one space between functions.